### ISSUE TITLE
Feedback Request for Bedrock Configurator

### ISSUE THREAD

**Issue by pmac** [mozmar/www-config#1]
Hello all,

In this repo is the new configurator for bedrock. It will handle adding or changing any configurations you might need for any bedrock app you might want (including your demos if you'd like). Please check out [the docs](https://mozmar.github.io/www-config/) and any of the code in here and let me know what you think in the comments of this issue.

Thanks in advance!

/cc @jgmize @glogiotatidis @metadave @jpetto @alexgibson @craigcook

**Comment by alexgibson** [comment]
I very much like the overall approach, especially it can now be code reviewed 💯. Hopefully this will make it easier for us to maintain switches. Looking at the configs in this repo, there is already much that is out of date / no longer used, and also highlights some differences between stage and prod. This would probably be a good opportunity to tidy up so we have a clean slate again.

I also really like that we have an easily view-able (read sharable) link that shows all current switch configs: https://mozmar.github.io/www-config/configs/

A couple of questions for clarification:

- How do we remove switches? Is removing them from the config file all that's required? (would this be worth adding to the docs)
- For our own individual demo instances, we simply submit a PR with our own configs? (I can't see this happening much, only to test things like stub attribution, FxA iframes etc).

**Comment by pmac** [comment]
Thanks for the thoughtful response Alex! I agree that a primary benefit is the always-up-to-date list of these settings, and I'm actually quite fond of the color coding I did for boolean values :)

You've hit on the sticking points. I'm not sure how best to proceed with the ability to delete config keys. I have  2 ideas:

1. Add a separate file that lists keys that should be removed
2. Do a diff of switches (only switches)  in the repo and those in the app and remove those that don't exist in the repo but do in the app. 

I think 2 is probably the best and most intuitive way to go, but it could result in some unintended deletions should someone skip the process. 2 also doesn't account for deleting any other kind of setting because we can't do a diff of all settings since the list in the repo isn't exhaustive (there are other settings that rarely change, as well as secrets like `SECRET_KEY` and `DATABASE_URL`). I think that might be fine though as deleting other config keys is pretty rare and can be handled manually. I'd love your thoughts on this though.

**Comment by metadave** [comment]
this is great. It would be interesting to be able to toggle set differences between dev/stage/prod to see what's unset/set/typo'd per env.

**Comment by pmac** [comment]
Not sure what you mean @metadave. Do you mean an easy way to see a diff between the app configs in the repo?

**Comment by metadave** [comment]
for example: what vars are set in prod that aren't set in stage.

**Comment by pmac** [comment]
Yeah. I was wondering whether separating stage and prod is worthwhile for these kinds of settings, or if switches should mostly go in global.env and be pushed out sequentially to dev -> stage -> prod.

**Comment by metadave** [comment]
another idea is to have a page that shows a grid of all settings for dev/stage/prod, then you can eyeball which settings are missing/typo'd etc.

**Comment by alexgibson** [comment]
In keeping parity for stage/prod, I could see benefits of automating them to both have the same vars. There are instances where we do want to flip things first in stage, before going onto prod however (e.g. sha-1 bouncer, stub attribution etc). How would we manage these instances?

**Comment by pmac** [comment]
@alexgibson you'd still be able to update whichever app you wanted, it would just use the same config. So you could update the `global.env` (or perhaps it should be renamed to `bedrock-all.env`) and then push the `bedrock-dev` branch, and check it, then the `bedrock-stage` branch, and check that, all before pushing `bedrock-prod`. It would allow you the flexibility of when they went live, but less for what went live. You'd need in your PR to provide a good reason I think to make them inconsistent. The more I talk about this the more I like it actually. Though this may need to be kept as only for stage and prod, as dev is special b/c of the `DEV=True` setting.

**Comment by alexgibson** [comment]
Ah that makes sense, thanks for the explanation pmac 👍

**Comment by jpetto** [comment]
Love that this will enable code review on config changes. No more inspecting my terminal command for typos with bomb-squad level deliberation.

In terms of deleting settings, since the impact could potentially be very  high, I think I'd prefer something more explicit than excluding the to-be-removed settings in the changed file. A specific file containing settings to be removed would be more easily reviewed, both in the moment and historically.

How much of a time impact would we be talking about in regards to getting settings live? If using the `fast` option, how much slower would it be to change production settings? I can foresee some emergency where a switch needs to be turned off *OMGRIGHTNOW*. Assuming that could be done manually, but what if someone with credentials is unavailable?

+1 on the suggestion to create a grid of all settings and values for all (major) environments.

**Comment by pmac** [comment]
> A specific file containing settings to be removed would be more easily reviewed, both in the moment and historically.

I agree and disagree. I think the review experience would be good, but possibly not better since you can very easily see a removal in a diff, and get the full history of changes in git of a single file. But I also think it is more explicit in a separate file. However, I can foresee some conflicts that could be non-obvious since you could add a setting to the delete file but fail to check the main file, and then how would it act if a variable was in both?

I've had an idea for a 3rd option... what if there were a value you could assign to a variable that meant "DELETE ME" to the system. So the variable could stay in the same file (to make it most obvious what the full state should be) but it would also be obvious which vars should not be set. I'm thinking something like:

```shell
REGULAR_VAR=true

# switches
SWITCH_DUDE=abiding

# deleted
SWITCH_WALTER=
```

I also thought it could be `SWITCH_WALTER=__DELETED__` or something, but in the shell setting a variable to nothing like that does kindof delete it. And it should be easily discoverable that's the deal.

I also still think it'd be a fine thing to just do a full diff on `SWITCH_*` var names and delete the ones that are no longer in the file in the repo, but I don't disagree that it's less explicit and possibly too magic.

What say you all?

**Comment by jpetto** [comment]
> what if there were a value you could assign to a variable that meant "DELETE ME" to the system

I like this line of thinking. What if we prefixed the line with something that would signal a deletion? A bang seems like the obvious choice...

```
SWITCH_DUDE=abiding
!SWITCH_WALTER=calmerthanyouare
SWITCH_DONNIE=outofhiselement
```

Would this make parsing the file more difficult? Or perhaps simpler?

**Comment by pmac** [comment]
> Would this make parsing the file more difficult?

That is the issue. The parser validates the key values and special characters aren't allowed. The easiest thing would be a value that is either blank or specially formatted. Blank seems most obvious and least error prone to me. I also like the idea of moving the deleted ones to a group at the bottom of the file.

**Comment by jpetto** [comment]
> I also like the idea of moving the deleted ones to a group at the bottom of the file.

👍

**Comment by pmac** [comment]
I just added the ability to delete variables by setting it to a blank value. [This is documented](https://mozmar.github.io/www-config/#editing-the-configs) now as well.

The change also improves the configurator by relying on a diff of the config you'd like (in the repo) and what is live in the app, and only setting (or deleting) those variables that differ. This should make it more efficient and the logs more useful.
